In [236]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [237]:
# Turning project folder path into a string
path=r'/Users/jasmeetakaur/Downloads/Jasmeeta/Data Immersion Course/Achievement 1/Achievement 4/Instacart Basket Analysis'

In [238]:
# Importing the “orders.csv” file
df_ords=pd.read_csv(os.path.join(path,'02 Data/Original Data/orders.csv'), index_col=False)

In [239]:
# Importing the products.csv file
df_prods=pd.read_csv(os.path.join(path,'02 Data/Original Data/products.csv'), index_col=False)

In [240]:
# Importing
df_dep=pd.read_csv(os.path.join(path,'02 Data/Original Data/departments.csv'), index_col=False)

In [241]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [242]:
df_dep_t = df_dep.T

In [243]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [244]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [245]:
new_header=df_dep_t.iloc[0]

In [246]:
df_dep_t_new=df_dep_t[1:]

In [247]:
df_dep_t_new.columns=new_header

In [248]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [268]:
df_ords=df_ords.drop(columns=['eval_set'])

In [269]:
df_ords.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329       1             1                  2                  8   
1   2398795       1             2                  3                  7   
2    473747       1             3                  3                 12   
3   2254736       1             4                  4                  7   
4    431534       1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [251]:
# changing the column name orders_dow to orders_day_of_week
df_ords.rename(columns={'order_dow':'order_day_of_week'},inplace=True)
df_ords

order_id  user_id eval_set  order_number  order_day_of_week  \
0         2539329        1    prior             1                  2   
1         2398795        1    prior             2                  3   
2          473747        1    prior             3                  3   
3         2254736        1    prior             4                  4   
4          431534        1    prior             5                  4   
...           ...      ...      ...           ...                ...   
3421078   2266710   206209    prior            10                  5   
3421079   1854736   206209    prior            11                  4   
3421080    626363   206209    prior            12                  1   
3421081   2977660   206209    prior            13                  1   
3421082    272231   206209    train            14                  6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

# Wrangling procedures for the task-4.4

In [270]:
df_ords.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329       1             1                  2                  8   
1   2398795       1             2                  3                  7   
2    473747       1             3                  3                 12   
3   2254736       1             4                  4                  7   
4    431534       1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [271]:
#3. Changing the name of column 'days_since_prior_order' to 'days_since_purchase' without overwriting the dataframe
# The changed column is only displayed below. Original df_ords is not modified.
df_ords.rename(columns={'days_since_prior_order':'days_since_purchase'},inplace=False)

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0         2539329       1             1                  2                  8   
1         2398795       1             2                  3                  7   
2          473747       1             3                  3                 12   
3         2254736       1             4                  4                  7   
4          431534       1             5                  4                 15   
...           ...     ...           ...                ...                ...   
3421078   2266710  206209            10                  5                 18   
3421079   1854736  206209            11                  4                 10   
3421080    626363  206209            12                  1                 12   
3421081   2977660  206209            13                  1                 12   
3421082    272231  206209            14                  6                 14   

         days_since_purchase  
0                        NaN  
1                       15.0  
2                       21.0  
3                       29.0  
4                       28.0  
...                      ...  
3421078                 29.0  
3421079                 30.0  
3421080                 18.0  
3421081                  7.0  
3421082                 30.0  

[3421083 rows x 6 columns]

In [272]:
# The following confirms that the df_ords still has original column name.
df_ords.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329       1             1                  2                  8   
1   2398795       1             2                  3                  7   
2    473747       1             3                  3                 12   
3   2254736       1             4                  4                  7   
4    431534       1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [255]:
# 2. Changing the data type of user_id to string
# Current type of user_id 
df_ords['user_id'].dtype

dtype('int64')

In [256]:
# Modifying the data type of user_id to string
df_ords['user_id']=df_ords['user_id'].astype('str')                                       

In [257]:
# New type of user_id
df_ords['user_id'].dtype

dtype('O')

In [258]:
# 4. To find the busiest hour for placing orders
df_ords['order_hour_of_day'].value_counts().head(1)

order_hour_of_day
10    288418
Name: count, dtype: int64

#  busiest hour for placing orders: 10 A.M. that has 288418 counts
 

In [259]:
# Creating data dictionary of department data frame
data_dict_t = df_dep_t_new.to_dict('index')

In [260]:
data_dict_t

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [261]:
# 5 Determining the meaning behind value of 4 in the "department_id" column
data_dict_t.get('4')

{'department': 'produce'}

In [262]:
# 6. Subset containing information about breakfast item sales
df_breakfast =  df_prods[df_prods['department_id']==14]

In [263]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [264]:
# 7. Every information people will need to throw dinner parties (include items from the following departments: alcohol, deli, beverages, and meat/seafood)
df_dinner = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]
df_dinner

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

# 8. The last dataframe created is df_dinner which has 7650 rows

In [265]:
# 9. Information about user with user_id=1
df_user1=df_ords[df_ords['user_id']=='1']
df_user1

order_id user_id eval_set  order_number  order_day_of_week  \
0    2539329       1    prior             1                  2   
1    2398795       1    prior             2                  3   
2     473747       1    prior             3                  3   
3    2254736       1    prior             4                  4   
4     431534       1    prior             5                  4   
5    3367565       1    prior             6                  2   
6     550135       1    prior             7                  1   
7    3108588       1    prior             8                  1   
8    2295261       1    prior             9                  1   
9    2550362       1    prior            10                  4   
10   1187899       1    train            11                  4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

# 10. Some details about the user with user_id=1 behavior from the stats we have:
1. This user only orders during the first four days of the week and never on days 5,6 and 7
2. This user doesn't order during late evening after 4 pm.

In [273]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [267]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))